File: thanh123456_AcceptedCodeVNSPOJ/NKINV.pas
const
    fin='';fon='';
    maxn=60000;
type
    mang=array [0..maxn*100] of longint;
var
    n,m:longint;
    a,it,pos:mang;
procedure nhap;
    var
        i:longint;
    begin
        assign(input,fin);reset(input);
        assign(output,fon);rewrite(output);
        readln(n);
        m:=0;
        for i:=1 to n do
            begin
                readln(a[i]);
                if a[i]>m then m:=a[i];
            end;
    end;
procedure build(i,l,r:longint);
    var
        mid:longint;
    begin
        if l=r then
            begin
                pos[l]:=i;
                exit;
            end;
        mid:=(l+r) div 2;
        build(i*2,l,mid);
        build(i*2+1,mid+1,r);
    end;
procedure UpDate(x:longint);
    var
        i:longint;
    begin
        i:=pos[x];
        while i>0 do
            begin
                inc(it[i]);
                i:=i div 2;
            end;
    end;
function get(i,l,r,u,v:longint):longint;
    var
        mid,left,right:longint;
    begin
        if (r<u) or (l>v) then exit(0)
            else
                if (l>=u) and (r<=v) then exit(it[i])
                    else
                        begin
                            mid:=(l+r) div 2;
                            left:=get(i*2,l,mid,u,v);
                            right:=get(i*2+1,mid+1,r,u,v);
                            exit(left+right);
                        end;
    end;
procedure solve;
    var
        i,res:longint;
    begin
        res:=0;
        for i:=n downto 1 do
            begin
                res:=res+get(1,1,m,1,a[i]-1);
                UpDate(a[i]);
            end;
        writeln(res);
        close(input); close(output);
    end;
begin
    nhap;
    build(1,1,m);
    solve;
end.

File: thanh123456_AcceptedCodeVNSPOJ/NKINV.cpp
#include <bits/stdc++.h>
#define pb push_back
#define p push
#define mp make_pair
using namespace std;
typedef pair<int,int> i2;
const int N = 6e4;
const int oo = 1e9 + 100;
int n,it[N * 4 + 1];
void nhap()
{
    scanf("%d",&n);
}
void update(int i,int l,int r,int u)
{
    if (l > u || r < u) {return;}
    it[i]++;
    if (l == r) {return;}
    int mid = (l + r) >> 1;
    update(i << 1,l,mid,u);
    update(i << 1 | 1,mid + 1,r,u);
}
int get(int i,int l,int r,int u,int v)
{
    if (l > v || r < u) {return 0;}
    if (l >= u && r <= v) {
        return it[i];
    }
    int mid = (l + r) >> 1;
    int left = get(i << 1,l,mid,u,v);
    int right = get(i << 1 | 1,mid + 1,r,u,v);
    return left + right;
}
void solve()
{
    long long res = 0;
    for (int i = 1; i <= n; i++) {
        int x;
        scanf("%d",&x);
        res += get(1,1,N,x + 1,N);
        update(1,1,N,x);
    }
    cout<<res;
}
int main()
{
    nhap();
    solve();
}
